packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts

```diff
@@ -1015,11 +1018,25 @@ export const model = (currentState: State, resW: ResponseType<AllActionStates>):
     if (Either.isRight(res)) {
       return {
         ...stateP,
-        controlState: 'UPDATE_TARGET_MAPPINGS',
+        controlState: 'COMPARE_MAPPINGS',
       };
     } else {
       throwBadResponse(stateP, res);
     }
+  } else if (stateP.controlState === 'COMPARE_MAPPINGS') {
+    const res = resW as ResponseType<typeof stateP.controlState>;
+    if (Either.isLeft(res) || !res.right.match) {
+      return {
+        ...stateP,
+        controlState: 'UPDATE_TARGET_MAPPINGS',
+      };
+    }
+
+    // The md5 of the mappings match, so there's no need to update target mappings
+    return {
```

Comment:
I need to think through this more carefully over a cup of coffee. But if we're just wanting to skip the `UPDATE_TARGET_MAPPINGS` then we probably need to have the same logic as in `UPDATE_TARGET_MAPPINGS_WAIT_FOR_TASK` where we could either go to `DONE` or `MARK_VERSION_INDEX_READY`

https://github.com/elastic/kibana/blob/84d3f3857da0654031652ffca8cdc0d90877eb94/packages/core/saved-objects/core-saved-objects-migration-server-internal/src/model/model.ts#L1037-L1055

Reply:
There's actually a race condition I didn't previously consider. One node could complete the `UPDATE_TARGET_MAPPINGS` step which updates the `_meta` of the mappings, but this node waits for a long time for the `UPDATE_TARGET_MAPPINGS_WAIT_FOR_TASK`.

In the meantime another node starts up, sees a version index, does the `COMPARE_MAPPINGS` and decides these indices are already migrated we can go to `DONE`/`MARK_VERSION_INDEX_READY`. 

But if this second node starts querying it won't find all results because `UPDATE_TARGET_MAPPINGS_WAIT_FOR_TASK` is still taking e.g. 3 mins to complete.

To fix this we probably need to add another mappings step:
 1. `COMPARE_MAPPINGS` (what about calling this step `COMPARE_TARGET_MAPPINGS` `CHECK_TARGET_MAPPINGS`)
 2. `UPDATE_TARGET_MAPPINGS`
 3. `UPDATE_TARGET_MAPPINGS_WAIT_FOR_TASK`
 4. `UPDATE_TARGET_MAPPINGS_META` a node will only complete this step after (3) completes

Reply:
Thanks a lot for all the insight, it’s really helpful!

I see your point, I imagine other Kibana nodes cannot simply sync on the `UPDATE_TARGET_MAPPINGS_WAIT_FOR_TASK` as they don’t have the `taskId` in their _state_. 

It seems a bit weird that multiple nodes have to compete with each other to run the different migration steps:

What happens if 2 Kibana nodes are started simultaneously, and they both run (1, 2, 3, 4)...

- Will they both run the will the costly `updateAndPickupMappings` method?
- Will the second override the first’s updates?
  - What if the first has more plugins enabled than the second?

Reply:
> Will they both run the will the costly updateAndPickupMappings method?
Yes
> Will the second override the first’s updates?
It depends a bit. updateByQuery uses optimistic concurrency control so if both run exactly at the same time then there will be version conflicts and only one write will succeed. But if they run slightly out of step then there will be N writes to each document (where N is the number of Kibana nodes). Yes, this is potentially quite expensive.
> What if the first has more plugins enabled than the second?
You're right, there's a race condition here where the Node with the new plugin gets update conflicts so these documents don't get their new mappings fields picked up. But if a kibana index has documents belonging to an unknown plugin we will refuse to start. So these two nodes can only "race" if there's no documents, but if there's no documents then it doesn't matter cause there's nothing to pickup. But the original migration shipped with this race condition and it was a coincidence that we fixed it, so well spotted!